### PR TITLE
Add reaction contract and link constrains to it

### PR DESCRIPTION
**Pull request by inwenc**
Add new link constraints between message-reactions and user

Change-type: minor
Signed-off-by: YingwenChen[yingwenchen@balena.io](mailto:yingwenchen@balena.io)

### CODE REVIEW

File: lib/link-constraints.ts [hunk 1]
```diff
@@ -1716,6 +1716,46 @@ export const constraints: LinkConstraint[] = [
 			inverse: "link-constraint-message-reaction-is-made-by-user",
 		},
 	},
+	{
+		slug: "link-contrain-relationship-reaction-is-attached-to-message",
```
Comment (joshbwlng) [inline lib/link-constraints.ts:1720]: Typo: `constrain` should be `constraint`. Also see it below.

Comment (joshbwlng) [reply lib/link-constraints.ts:1720]: Also don't need `relationship` in the slug

File: lib/link-constraints.ts [hunk 2]
```diff
@@ -1716,6 +1716,46 @@ export const constraints: LinkConstraint[] = [
...
+		name: "has attached element",
+		data: {
+			title: "Reaction contract",
+			from: "message",
+			to: "reaction",
+			inverse: "link-contrain-relationship-reaction-is-attached-to-message",
+		},
+	},
+	{
+		slug: "link-constrain-relationship-reaction-is-attached-to-whisper",
```
Comment (joshbwlng) [inline lib/link-constraints.ts:1740]: Don't need `relationship` in the slug

File: lib/link-constraints.ts [hunk 3]
```diff
@@ -1716,6 +1716,46 @@ export const constraints: LinkConstraint[] = [
...
+		},
+	},
+	{
+		slug: "link-constrain-Whisper",
+		name: "is attached to",
+		data: {
+			title: "Reaction",
+			from: "whisper",
+			to: "reaction",
+			inverse: "link-constrain-relationship-reaction-is-attached-to-whisper",
```
Comment (joshbwlng) [inline lib/link-constraints.ts:1756]: Don't need `relationship` here either

File: lib/link-constraints.ts [hunk 4]
```diff
@@ -1716,6 +1716,46 @@ export const constraints: LinkConstraint[] = [
...
+		name: "is attached to",
+		data: {
+			title: "Reaction",
+			from: "reaction",
+			to: "whisper",
+			inverse: "link-constrain-Whisper",
+		},
+	},
+	{
+		slug: "link-constrain-Whisper",
```
Comment (joshbwlng) [inline lib/link-constraints.ts:1750]: Incomplete slug

**Comment by joshbwlng** [comment]
@inwenc Looks like your prettier plugin has autoformatted everything again. Can you fix that and then I'll review the changes again.

**Comment by inwenc** [comment]
@joshbwlng - but by looking just the changes of my commits after your push, they don't look like I applied prettier, again.

**Comment by joshbwlng** [comment]
@inwenc Thanks for cleaning it up

File: lib/link-constraints.ts [hunk 5]
```diff
@@ -1676,6 +1696,66 @@ export const constraints: LinkConstraint[] = [
...
+	{
+		slug: 'link-constraint-user-react-to-message',
+		name: 'reacted',
+		data: {
+			title: 'Reaction contract',
+			from: 'user',
+			to: '*',
+			inverse: 'link-constraint-message-reaction-is-made-by-user',
+		},
+	},
```
Comment (joshbwlng) [inline lib/link-constraints.ts:1718]: Don't think we need this either

File: lib/link-constraints.ts [hunk 6]
```diff
@@ -1676,6 +1696,66 @@ export const constraints: LinkConstraint[] = [
...
+	{
+		slug: 'link-contraint-reaction-is-attached-to-message',
+		name: 'is attached to',
+		data: {
+			title: 'Reaction',
+			from: 'reaction',
+			to: 'message',
+			inverse: 'link-constraint-has-attached-element',
```
Comment (joshbwlng) [inline lib/link-constraints.ts:1726]:
```suggestion
			inverse: 'link-constraint-message-has-attached-element-reaction',
```

File: lib/link-constraints.ts [hunk 7]
```diff
@@ -1676,6 +1696,66 @@ export const constraints: LinkConstraint[] = [
...
+		},
+	},
+	{
+		slug: 'link-constraint-has-attached-element',
```
Comment (joshbwlng) [inline lib/link-constraints.ts:1730]:
```suggestion
		slug: 'link-constraint-message-has-attached-element-reaction',
```

File: lib/link-constraints.ts [hunk 8]
```diff
@@ -1676,6 +1696,66 @@ export const constraints: LinkConstraint[] = [
...
+		},
+	},
+	{
+		slug: 'link-constraint-reaction-is-attached-to-whisper',
+		name: 'is attached to',
+		data: {
+			title: 'Reaction',
+			from: 'reaction',
+			to: 'whisper',
+			inverse: 'link-constraint-whisper-has-attached-reaction',
```
Comment (joshbwlng) [inline lib/link-constraints.ts:1746]:
```suggestion
			inverse: 'link-constraint-whisper-has-attached-element-reaction',
```

File: lib/link-constraints.ts [hunk 9]
```diff
@@ -1676,6 +1696,66 @@ export const constraints: LinkConstraint[] = [
...
+		},
+	},
+	{
+		slug: 'link-constraint-whisper-has-attached-reaction',
```
Comment (joshbwlng) [inline lib/link-constraints.ts:1750]:
```suggestion
		slug: 'link-constraint-whisper-has-attached-element-reaction',
```

**Comment by joshbwlng** [comment]
@inwenc There are still a few unresolved suggested changes

File: lib/link-constraints.ts [hunk 10]
```diff
@@ -1676,6 +1676,46 @@ export const constraints: LinkConstraint[] = [
...
+			from: 'reaction',
+			to: 'message',
+			inverse: 'link-constraint-message-has-attached-element-reaction',
+		},
+	},
+	{
+		slug: 'link-constraint-message-has-attached-element-reaction',
+		name: 'has attached element',
+		data: {
+			title: 'Reaction contract',
```
Comment (joshbwlng) [inline lib/link-constraints.ts:1693]:
```suggestion
			title: 'Reaction',
```

Nit pick, but there's no reason this one should have a different title than others

File: lib/link-constraints.ts [hunk 11]
```diff
@@ -1676,6 +1676,46 @@ export const constraints: LinkConstraint[] = [
 			inverse: 'link-constraint-loop-has-product-repository',
 		},
 	},
+	{
+		slug: 'link-contraint-reaction-is-attached-to-message',
+		name: 'is attached to',
+		data: {
+			title: 'Reaction',
```
Comment (joshbwlng) [inline lib/link-constraints.ts:1683]:
```suggestion
			title: 'Message',
```

Just following the pattern of other link constraint definitions

File: lib/link-constraints.ts [hunk 12]
```diff
@@ -1676,6 +1676,46 @@ export const constraints: LinkConstraint[] = [
...
+			from: 'message',
+			to: 'reaction',
+			inverse: 'link-contraint-reaction-is-attached-to-message',
+		},
+	},
+	{
+		slug: 'link-constraint-reaction-is-attached-to-whisper',
+		name: 'is attached to',
+		data: {
+			title: 'Reaction',
```
Comment (joshbwlng) [inline lib/link-constraints.ts:1703]:
```suggestion
			title: 'Whisper',
```

**Review by joshbwlng — APPROVED** [review]

**Comment by joshbwlng** [comment]
@balena-ci rebase

**Comment by inwenc** [comment]
@balena-ci rebase

**Pull request merged**